### ISSUE TITLE
Nullable: Cannot use a type without a warning - constraint always fails

### ISSUE THREAD

**Issue by krwq** [dotnet/roslyn#35406]
**Version Used**: 
sharplab with master (Apr 30)

**Steps to Reproduce**:

```csharp
#nullable enable
using System;

public interface IEquatable2<T>
{
  bool Equals(T a, T b);
}

public class Foo: IEquatable2<Foo?>
{
  public bool Equals(Foo? a, Foo? b)
  {
    return true;
  }
}

public class Bar<T> where T : IEquatable2<T>
{
}

public class C {
    public void M() {
      Type a = typeof(Bar<Foo>);
      Type b = typeof(Bar<Foo?>);
    }
}
```

**Expected Behavior**:

No warnings.

**Actual Behavior**:

2 Warnings:

```
    warning CS8631: The type 'Foo' cannot be used as type parameter 'T' in the generic type or method 'Bar<T>'. Nullability of type argument 'Foo' doesn't match constraint type 'IEquatable2<Foo>'.
    warning CS8631: The type 'Foo?' cannot be used as type parameter 'T' in the generic type or method 'Bar<T>'. Nullability of type argument 'Foo?' doesn't match constraint type 'IEquatable2<Foo?>'.
```

Sorry for the title but it's hard to write a sentence from above repro

cc: @dotnet/nullablefc

**Comment by jcouv** [comment]
I think the current warnings are correct. 
Converting `Foo` (which is an `IEquatable2<Foo?>`) to `IEquatable2<Foo>` should warn (because of nested nullability).
Converting `Foo?` to `IEquatable<Foo?>` also should warn (because of top-level nullability).

In this scenario, the way out is to declare `interface IEquatable2<out T>` instead. Then `Bar<Foo>` is okay ([sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABADAAmwEYBuAWAChKBiAOwFcAbRiYRufOW19yy7AMz4AlrXgIAZhADGHAJIBRAI70IMHnABMAHlH4AKgD5KAb0r4L+YAHtrjfMtWMoACn34ISA1YCU5CgC+fBSChJr4AGK2IPiKKmoaOlHWAPzGFGYUloRCNnYO8c4uySkeXiW+5paZ2dnYAOz4MAj0cP7ZQYHBodjhAEIQCNpG+ADuABaIHO4xcarqbFrD6Zmd/EK9+ADC+KZVFj0ALPgAsi4+u1nV+7X6AJ4ADhwQ+AC8TY9w1hIuA0PJhj8N06ASAA=)).
